xapian-letor/api/featurevector_internal.h

```diff
@@ -0,0 +1,76 @@
+/* featurevector.cc: The file responsible for transforming the document into the feature space.
+ *
+ * Copyright (C) 2012 Parth Gupta
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#ifndef FEATURE_VECTOR_INTERNAL_H
+#define FEATURE_VECTOR_INTERNAL_H
+
+#include <config.h>
+
+#include <xapian.h>
+
+#include "xapian-letor/featurevector.h"
+#include "xapian-letor/featuremanager.h"
+
+#include <list>
+#include <map>
+
+#include "str.h"
+#include "stringutils.h"
+
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "safeerrno.h"
+#include "safeunistd.h"
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+
+
+using namespace std;
+
+namespace Xapian{
+
+class FeatureVector::Internal : public Xapian::Internal::intrusive_base
+{
+    friend class FeatureVector;
+    double label;
+    double score;
+    std::map<int,double> fvals;
```

Comment:
Another approach to this is to make the `FeatureVector` into an actual vector of feature values, since (once calculated for each document) it doesn't actually matter how they were calculated, just that the vectors are the same in the system. That would allow for more flexibility by users, who could configure the system (from training through to use) with the feature generators they actually want to use. (With a default set that is our 'recommended' approach.) One of the big advantages there is that we wouldn't have to assign numbers universally, so we wouldn't have to worry about where to put experimental features, for instance.

Alternatively, if we do need ids for them, we could use URIs and have a registry which converts them to numbers (somehow). That again would be within the system.

Hopefully as we progress across the summer it'll become clear which route is more appropriate. If we do continue using identifiers that are numeric at a low level, then definitely using an `enum` type would be an improvement.